File: Data Structures & Algorithms/top-k-elements-in-list/submission-1.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        int n = nums.size();
        unordered_map<int, int> counts;
        for (int num : nums) {
            counts[num]++;
        }

        // Buckets where index is the frequency
        // We need n + 1 size because a number could appear n times
        vector<vector<int>> buckets(n + 1);
        for (auto const& [val, freq] : counts) {
            buckets[freq].push_back(val);
        }

        vector<int> result;
        // Search from highest frequency to lowest
        for (int i = n; i >= 0 && result.size() < k; i--) {
            for (int val : buckets[i]) {
                result.push_back(val);
                if (result.size() == k) return result;
            }
        }
        return result;
    }
};

File: Data Structures & Algorithms/binary-search/submission-1.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        low, s = 0, len(nums)
        high = s - 1

        while low <= high:
            mid = (low + high) // 2
            if (nums[mid] == target):
                return mid
            elif nums[mid] < target:
                low = mid + 1
            else:
                high = mid - 1
        return -1
        



"""

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int low = 0;
        int s = nums.size();
        int high = s - 1;
        while (low <= high) {
            int mid = (low + high)/2;
            if (nums[mid] == target) {
                return mid;
            } else if (nums[mid] < target) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }

        }
        return -1;
    }
};

"""